types/lib/mapper.go

```diff
@@ -0,0 +1,276 @@
+
+	Delete(sdk.Context, uint64)
+
+	Push(sdk.Context, interface{})
+
+	// Getter/Setter for meta information - can be customized
+	GetMeta(sdk.Context, interface{}) error
+	SetMeta(sdk.Context, interface{})
+
+	// Iterate*() is used to iterate over all existing elements in the list
+	// Return true in the continuation to break
+
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	IterateRead(sdk.Context, interface{}, func(sdk.Context, uint64) bool)
+
+	// IterateWrite() is safe to write over the domain
+	IterateWrite(sdk.Context, interface{}, func(sdk.Context, uint64) bool)
+
+	// Key for the length of the list
+	LengthKey() []byte
+
+	// Key for getting elements
+	ElemKey(uint64) []byte
+
+	// Key for additional meta information of the list
+	MetaKey() []byte
+}
+
+func NewListMapper(cdc *wire.Codec, key sdk.StoreKey, prefix string) ListMapper {
+	return GenericMapper{
+		key:    key,
+		cdc:    cdc,
+		prefix: prefix,
+	}
+}
+
+func (lm GenericMapper) Len(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.LengthKey())
+	if bz == nil {
+		zero, err := lm.cdc.MarshalBinary(0)
+		if err != nil {
+			panic(err)
+		}
+		store.Set(lm.LengthKey(), zero)
+		return 0
+	}
+	var res uint64
+	if err := lm.cdc.UnmarshalBinary(bz, &res); err != nil {
+		panic(err)
+	}
+	return res
+}
+
+func (lm GenericMapper) Get(ctx sdk.Context, index uint64, ptr interface{}) error {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.ElemKey(index))
+	return lm.cdc.UnmarshalBinary(bz, ptr)
+}
+
+func (lm GenericMapper) Set(ctx sdk.Context, index uint64, value interface{}) {
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.ElemKey(index), bz)
+}
+
+func (lm GenericMapper) Delete(ctx sdk.Context, index uint64) {
+	store := ctx.KVStore(lm.key)
+	store.Delete(lm.ElemKey(index))
+}
+
+func (lm GenericMapper) Push(ctx sdk.Context, value interface{}) {
+	length := lm.Len(ctx)
+	lm.Set(ctx, length, value)
+
+	store := ctx.KVStore(lm.key)
+	store.Set(lm.LengthKey(), marshalUint64(lm.cdc, length+1))
+}
+
+func (lm GenericMapper) GetMeta(ctx sdk.Context, ptr interface{}) error {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.MetaKey())
+	return lm.cdc.UnmarshalBinary(bz, ptr)
+}
+
+func (lm GenericMapper) SetMeta(ctx sdk.Context, value interface{}) {
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.MetaKey(), bz)
+}
+
+func (lm GenericMapper) IterateRead(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, uint64) bool) {
+	store := ctx.KVStore(lm.key)
+	start, end := subspace([]byte(fmt.Sprintf("%s/elem/", lm.prefix)))
+	iter := store.Iterator(start, end)
+	for ; iter.Valid(); iter.Next() {
+		v := iter.Value()
+		if err := lm.cdc.UnmarshalBinary(v, ptr); err != nil {
+			panic(err)
+		}
+		s := strings.Split(string(iter.Key()), "/")
```

Comment:
Is this necessary? We should know the indices (starting at 0, ending at length of list - 1), no need to parse strings. Also, this will calculate the wrong indices if elements are deleted.